flow/probes/sflow.go

```diff
@@ -190,3 +183,136 @@ func NewSFlowProbe(a string, p int, g *graph.Graph) (*SFlowProbe, error) {
 
 	return probe, nil
 }
+
+type SkydivePacket struct {
+	gopacket.Packet
+	ci     gopacket.CaptureInfo
+	raw    *sflow.RawPacketFlow
+	sample *sflow.FlowSample
+}
+
+type SFlowReader struct {
+	gopacket.PacketSource
+	ProbeTarget      string
+	PacketChannel    chan SkydivePacket
+	RawPacketCounter uint64
+}
+
+func (r *SFlowReader) LinkType() layers.LinkType {
+	return layers.LinkTypeEthernet
+}
+
+func (r *SFlowReader) CreatePacket(sample *sflow.FlowSample, raw *sflow.RawPacketFlow) {
+	var ci gopacket.CaptureInfo
+	ci.Timestamp = time.Now()
+	ci.CaptureLength = int(raw.FrameLength)
+	ci.Length = int(raw.HeaderSize)
+
+	data := raw.Header
+	r.PacketChannel <- SkydivePacket{
+		gopacket.NewPacket(data, r.LinkType(), gopacket.DecodeOptions{}),
+		ci,
+		raw,
+		sample}
+}
+
+/* PacketSource */
+func (r *SFlowReader) NextPacket() (SkydivePacket, error) {
+	packet := <-r.PacketChannel
+	m := packet.Metadata()
+	m.CaptureInfo = packet.ci
+	m.Truncated = m.Truncated || packet.ci.CaptureLength < packet.ci.Length
+
+	return packet, nil
+}
+
+func (p *SFlowReader) packetsToChannel() {
+	defer close(p.PacketChannel)
+	for {
+		packet, err := p.NextPacket()
+		if err == io.EOF {
+			return
+		} else if err == nil {
+			p.PacketChannel <- packet
```

Comment:
Something I don't understand we are reading the PacketChannel thanks to the NextPacket function a then we re-inject to the same channel what we just read ? we will have a infinite loop, iterating over the same packet ?